### PR TITLE
Initial take at a native unit test harness.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Steps to add a unit test target:

* Create a target using the "test" template in "testing/test.gni"
* Add your test files as sources to this target.
* Add the "//sky:sky_test" target as a dep.
* Depend on the new target from somewhere (usually root BUILD.gn)

cc @jason-simmons, @abarth

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Only works on Mac for now. Wiring up the iOS, Android and Linux stuff next.

File: build/secondary/testing/gtest/BUILD.gn [hunk 1]
```diff
@@ -18,7 +18,10 @@ config("gtest_config") {
   ]
 
   # Gtest headers need to be able to find themselves.
-  include_dirs = [ "include" ]
+  include_dirs = [
```
Comment (jamesr) [inline build/secondary/testing/gtest/BUILD.gn]: why do you need this change?

Comment (chinmaygarde) [inline build/secondary/testing/gtest/BUILD.gn]: That would mean looking at "//build/secondary/testing/gtest/include". There is no such path. The gtest checked out by deps is in //testing/gtest

Comment (jamesr) [inline build/secondary/testing/gtest/BUILD.gn]: in general i think i'd prefer to keep this aligned with upstream if it's functional

Comment (jamesr) [inline build/secondary/testing/gtest/BUILD.gn]: > That would mean looking at "//build/secondary/testing/gtest/include". There is no such path. The gtest checked out by deps is in //testing/gtest

I don't think that's how build/secondary works - in GN paths inside the secondary directory should map to their non-secondary locations transparently.  Do you have a 'gn check' failure showing the tool looking in the wrong place?

**Comment by jamesr** [comment]
What platforms is sky shell intended to be functional for testing on?  I haven't looked in a while but shells on desktop platforms were pretty minimal.

**Comment by jamesr** [comment]
I think it'd also be useful to have an example test to see how it wires up and runs.

File: build/secondary/testing/gtest/BUILD.gn (hunk 1 above)
Comment (chinmaygarde) [inline build/secondary/testing/gtest/BUILD.gn]: Hmm, the compile error suggested it was looking in secondary. I will check again.

**Comment by chinmaygarde** [comment]
> What platforms is sky shell intended to be functional for testing on? I haven't looked in a while but shells on desktop platforms were pretty minimal.

As of this patch, it is Mac only. I am wiring up iOS right now. Linux should be straightforward then Android.

> I think it'd also be useful to have an example test to see how it wires up and runs.

On the Mac, its just an executable. On iOS, its a framework a generic tester loads and gets its results from for display and reporting.

**Comment by Hixie** [comment]
Yay testing!

**Comment by jamesr** [comment]
I'm patching this in to see what's up with the secondary stuff.  I think if we can figure this out everything else looks very nice.

**Comment by jamesr** [comment]
There are some existing host unit tests already:

sky_wtf_unittests
sky_platform_unittests

that seem to be functional even without this patch.

**Comment by chinmaygarde** [comment]
That was super perplexing. It turns out I was including the test_support dependency in the wrong place while development that caused the issues with the paths being relative to an incorrect location.

I have reverted that change to the gtest BUILD.gn. PTAL.

**Comment by jamesr** [comment]
Did you look at the sky_{wtf,platform}_uinttests targets?  How does this patch relate to those?

**Comment by jamesr** [comment]
More specifically this commit says it's introducing a native unit test harness but it appears there already are existing working native unit tests in the build so I'm confused about what this is adding.

**Comment by chinmaygarde** [comment]
> How does this patch relate to those?

I want to create a test suite that runs on the target instead of the host. Specifically for things like testing  platform service integration and writing tests for individual service implementations on various platforms. Right now, the native test targets generate test executables that can run on the host. There is no way for me to write unit tests for code that will be executed on iOS for example (for the various services).

It is possible there is a better way of approaching this. I am open to ideas. Just wanted to get the conversation started. Me not being able to write unit tests for iOS specific code is causing me a fair amount of stress.

**Comment by abarth** [comment]
Those other targets also link with lower levels of the system and don't include the bulk of the code that might be interesting to test.

**Comment by jamesr** [comment]
Totally with you on wanting to write tests for the target platform - what I'm trying to figure out is how this is different from the other ones.  Is the difference that this runs tests inside of SkyShell, which presumably on iOS/Android/etc will have lots of additional supporting infrastructure?  If so then this seems like a good way to do that although it's not obvious that is the intent from the commit message / comments.  We found in chromium that trying to run unit tests in the context of a large binary that included the world was a lot less useful than having unit tests for components that only linked in the bits that the component under test was using, but both forms had value in different situations.

Regardless this seems like a perfectly good starting point for whole-shell tests if that's the intent.

**Comment by jamesr** [comment]
https://media.giphy.com/media/3oEduJ3b3ehucahuko/giphy.gif

**Comment by Hixie** [comment]
@chinmaygarde What's the status of this PR?

**Comment by chinmaygarde** [comment]
The stuff I primarily wanted to test with this is going away. Was waiting to add other tests first. Resuming.

**Pull request closed without merging**